Source/AMRInterpolator/SphericalGeometry.hpp

```diff
@@ -22,73 +24,121 @@
-    inline double get_grid_coord(int a_dir, double a_radius, double a_theta,
-                                 double a_phi) const
+    ALWAYS_INLINE double get_grid_coord(int a_dir, double a_radius,
+                                        double a_theta, double a_phi = 0.) const
     {
-        switch (a_dir)
+        double center = (a_dir < CH_SPACEDIM ? m_center[a_dir] : 0.);
+        // flip 'a_dir' such that the old 'z' points to 'm_up_dir'
+        switch ((a_dir + 2 - m_up_dir) % 3)
         {
         case (0):
-            return m_center[0] + a_radius * sin(a_theta) * cos(a_phi);
+            return center + a_radius * sin(a_theta) * cos(a_phi);
         case (1):
-            return m_center[1] + a_radius * sin(a_theta) * sin(a_phi);
+            return center + a_radius * sin(a_theta) * sin(a_phi);
         case (2):
-            return m_center[2] + a_radius * cos(a_theta);
+            return center + a_radius * cos(a_theta);
         default:
             MayDay::Error("SphericalGeometry: Direction not supported");
         }
     }
 
     //! returns the area element on a sphere with radius a_radius at the point
     //! (a_theta, a_phi)
-    inline double area_element(double a_radius, double a_theta,
-                               double a_phi) const
+    static ALWAYS_INLINE double area_element(double a_radius, double a_theta,
+                                             double a_phi)
     {
         return a_radius * a_radius * sin(a_theta);
     }
 
-    inline std::string param_name() const { return "r"; }
+    static ALWAYS_INLINE std::string param_name() { return "r"; }
+    static ALWAYS_INLINE std::string u_name() { return "theta"; }
+    static ALWAYS_INLINE std::string v_name() { return "phi"; }
 
-    inline std::string u_name() const { return "theta"; }
+    static const IntegrationMethod &
+    get_recommended_integration_method_u(int a_num_points_theta)
+    {
+        static const IntegrationMethod &simpson = IntegrationMethod::simpson;
+        static const IntegrationMethod &trapezium =
+            IntegrationMethod::trapezium;
+        if (simpson.is_valid(a_num_points_theta, is_u_periodic()))
+            return simpson;
+        MayDay::Warning("Use an odd number of 'u' points to use simpson rule. "
+                        "Defaulting to trapezium.");
+        return trapezium;
+    }
+
+    static const IntegrationMethod &
+    get_recommended_integration_method_v(int a_num_points_phi)
+    {
+        static const IntegrationMethod &simpson = IntegrationMethod::simpson;
+        static const IntegrationMethod &trapezium =
+            IntegrationMethod::trapezium;
+        if (simpson.is_valid(a_num_points_phi, is_v_periodic()))
+            return simpson;
+        MayDay::Warning("Use an even number of 'v' points to use simpson rule. "
+                        "Defaulting to trapezium.");
+        return trapezium;
+    }
 
-    inline std::string v_name() const { return "phi"; }
+  protected:
```

Comment:
I think you can make these `private` since we're not going to derive a child class from this class.